File: 74-SearchA2dMatrix/74-SearchA2dMatrix.java
// Last updated: 5/30/2025, 7:35:03 PM
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length, n = matrix[0].length;
        int top = 0, bot = m - 1; // Adjusted for rows
        while (top <= bot) {
            int mid = (top + bot) / 2;
            if (target > matrix[mid][n - 1]) {
                top = mid + 1;
            } else if (target < matrix[mid][0]) {
                bot = mid - 1;
            } else {
                break;
            }
        }
        if (top > bot) {
            return false;
        }
        int row = (top + bot) / 2;
        int left = 0, right = n - 1; // Renamed L and R for clarity
        while (left <= right) {
            int mid = (left + right) / 2;
            if (target < matrix[row][mid]) {
                right = mid - 1;
            } else if (matrix[row][mid] < target) {
                left = mid + 1;
            } else {
                return true;
            }
        }
        return false;
    }
}

File: 74-Searcha2DMatrix/74-Searcha2DMatrix.py
# Last updated: 5/31/2025, 6:38:09 PM
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        n = len(matrix)
        m = len(matrix[0])

        leftColumn = 0
        rightColumn = m - 1
        upRow = 0
        downRow = n - 1

        while upRow <= downRow:
            midRow = (upRow + downRow) // 2
            if target < matrix[midRow][leftColumn]:
                downRow = midRow - 1
            elif target > matrix[midRow][rightColumn]:
                upRow = midRow + 1
            else:
                while leftColumn <= rightColumn:
                    midColumn = (leftColumn+rightColumn) //2
                    if target < matrix[midRow][midColumn]:
                        rightColumn = midColumn -1
                    elif target > matrix[midRow][midColumn]:
                        leftColumn = midColumn + 1
                    else:
                        return True
        return False


